xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,607 @@
+
+        # clean up successor references to blocks for ops inside this region
+        for block, imm_block in zip(blocks, region.blocks):
+            dummy_block = block_map[block]
+            for block in region.blocks:
+                for op in block.ops:
+                    try:
+                        dummy_index = op.successors.index(dummy_block)
+                    except ValueError:
+                        continue
+                    # replace dummy successor with actual successor
+                    object.__setattr__(
+                        op, "successors",
+                        IList(op.successors[:dummy_index] + [imm_block] +
+                              op.successors[dummy_index + 1:]))
+
+        return region
+
+    def get_mutable_copy(
+            self,
+            value_mapping: dict[ISSAValue, SSAValue] | None = None,
+            block_mapping: dict[IBlock, Block] | None = None) -> Region:
+        """
+        Returns a mutable region that is a copy of this immutable region.
+        The value_mapping and block_mapping are used to map already known correspondings
+        of immutable values to mutable values and immutable blocks to mutable blocks.
+        """
+        if value_mapping is None:
+            value_mapping = {}
+        if block_mapping is None:
+            block_mapping = {}
+        mutable_blocks: list[Block] = []
+        # All mutable blocks have to be initialized first so that ops can
+        # refer to them in their successor lists.
+        for block in self.blocks:
+            mutable_blocks.append(
+                mutable_block := Block.from_arg_types(block.arg_types))
+            block_mapping[block] = mutable_block
+        for block in self.blocks:
+            # This will use the already created Block and populate it
+            block.get_mutable_copy(value_mapping=value_mapping,
+                                   block_mapping=block_mapping)
+        return Region.from_block_list(mutable_blocks)
+
+
+@dataclass(frozen=True)
+class IBlock:
+    """An immutable block contains a list of immutable operations. IBlocks are contained in IRegions."""
+    args: IList[IBlockArg]
+    ops: IList[IOperation]
+
+    @property
+    def arg_types(self) -> list[Attribute]:
+        frozen_arg_types = [arg.typ for arg in self.args]
+        return frozen_arg_types
+
+    def __hash__(self) -> int:
+        return (id(self))
+
+    def __eq__(self, __o: object) -> bool:
+        return self is __o
+
+    def __repr__(self) -> str:
+        return "block of" + str(len(self.ops)) + " with args: " + str(
+            self.args)
+
+    def __post_init__(self):
+        for arg in self.args:
+            object.__setattr__(arg, "block", self)
+
+    def __init__(self, args: Sequence[Attribute] | Sequence[IBlockArg],
+                 ops: Sequence[IOperation]):
+        """Creates a new immutable block."""
+
+        # Type Guards:
+        def is_iblock_arg_seq(
```

Comment:
I don't quite like the naming here. Why is it called `is...` and then returns the list? I understand that this is convenient to use, but I would personally name it something like `to_iblock_arg_seq` and then just return and empty list (or None) if the initial list was empty.

Considering that this is not part of our public API, this is not a very strong opinion. (though if we make it part of a utils file as Mathieu suggested, I think this should be renamed.)

Reply:
I think you got confused there, this returns a `TypeGuard` which is just a boolean ;)